pennylane/transforms/rz_phase_gradient.py

```diff
@@ -0,0 +1,192 @@
+            phase gradient wires, hence the precision of the angle :math:`\phi`.
+
+    Returns:
+        qnode (QNode) or quantum function (Callable) or tuple[List[QuantumTape], function]: The transformed circuit as described in :func:`qml.transform <pennylane.transform>`.
+
+    **Example**
+
+    .. code-block:: python
+
+        import pennylane as qml
+        import numpy as np
+        from functools import partial
+        from pennylane.transforms.rz_phase_gradient import rz_phase_gradient
+
+        precision = 3
+        phi = (1/2 + 1/4 + 1/8) * 2 * np.pi
+        wire="targ"
+        aux_wires = [f"aux_{i}" for i in range(precision)]
+        phase_grad_wires = [f"phg_{i}" for i in range(precision)]
+        work_wires = [f"work_{i}" for i in range(precision-1)]
+        wire_order = [wire] + aux_wires + phase_grad_wires + work_wires
+
+        def phase_gradient(wires):
+            # prepare phase gradient state
+            qml.X(wires[-1])
+            qml.QFT(wires)
+
+        @partial(qml.transforms.rz_phase_gradient, aux_wires=aux_wires, phase_grad_wires=phase_grad_wires, work_wires=work_wires)
+        @qml.qnode(qml.device("default.qubit"))
+        def rz_circ(phi, wire):
+            phase_gradient(phase_grad_wires) # prepare phase gradient state
+
+            qml.Hadamard(wire) # transform rotation
+            qml.RZ(phi, wire)
+            qml.Hadamard(wire) # transform rotation
+
+            return qml.probs(wire)
+
+    In this example we perform the rotation of an angle of :math:`\phi = (0.111)_2 2\pi`. Because phase shifts
+    are trivial on computational basis states, we transform the :math:`R_Z` rotation to `R_X = H R_Z H` via two
+    :class:`~.Hadamard` gates.
+
+    Note that for the transform to work, we need to also prepare a phase gradient state on the ``phase_grad_wires``.
+
+    Overall, the full circuit looks like the following:
+
+    >>> print(qml.draw(rz_circ, wire_order=wire_order)(phi, wire))
+      targ: ──H─╭●──────────────╭●───╭GlobalPhase(2.75)──H─┤  Probs
+     aux_0: ────├|Ψ⟩─╭SemiAdder─├|Ψ⟩─├GlobalPhase(2.75)────┤
+     aux_1: ────├|Ψ⟩─├SemiAdder─├|Ψ⟩─├GlobalPhase(2.75)────┤
+     aux_2: ────╰|Ψ⟩─├SemiAdder─╰|Ψ⟩─├GlobalPhase(2.75)────┤
+     phg_0: ────╭QFT─├SemiAdder──────├GlobalPhase(2.75)────┤
+     phg_1: ────├QFT─├SemiAdder──────├GlobalPhase(2.75)────┤
+     phg_2: ──X─╰QFT─├SemiAdder──────├GlobalPhase(2.75)────┤
+    work_0: ─────────├SemiAdder──────├GlobalPhase(2.75)────┤
+    work_1: ─────────╰SemiAdder──────╰GlobalPhase(2.75)────┤
+
+    The additional work wires are required by the :class:`~.SemiAdder`.
+    Executing the circuit, we get the expected result:
+
+    >>> rz_circ(phi, wire)
+    array([0.85355339, 0.14644661])
+
+    """
+    operations = []
+    for op in tape.operations:
+        if isinstance(op, qml.RZ):
+            operations.extend(
+                _rz_phase_gradient(
```

Comment:
I'm wondering whether it would be nicer to not queue one `GlobalPhase` per call to `_rz_phase_gradient`, but collect the total phase from all of the calls and queue a single `GlobalPhase`? 🤔